internal/storage/overlay/store.go

```diff
@@ -0,0 +1,213 @@
+func (s *Store) GetOverlayPolicyLoader(ctx context.Context, schemaMgr schema.Manager) (engine.PolicyLoader, error) {
+	getPolicyLoader := func(storeInterface storage.Store, key string) (engine.PolicyLoader, error) {
+		switch st := storeInterface.(type) {
+		case storage.SourceStore:
+			pl, err := compile.NewManager(ctx, st, schemaMgr)
+			if err != nil {
+				return nil, fmt.Errorf("failed to create %s compile manager: %w", key, err)
+			}
+			return pl, nil
+		case storage.BinaryStore:
+			return st, nil
+		default:
+			return nil, errors.New("overlaid store does not implement either SourceStore or BinaryStore interfaces")
+		}
+	}
+
+	var err error
+	if s.basePolicyLoader, err = getPolicyLoader(s.baseStore, "base"); err != nil {
+		return nil, err
+	}
+	if s.fallbackPolicyLoader, err = getPolicyLoader(s.fallbackStore, "fallback"); err != nil {
+		return nil, err
+	}
+
+	return s, nil
+}
+
+func withCircuitBreaker[T any](s *Store, baseFn, fallbackFn func() (T, error)) (T, error) {
+	if s.circuitBreaker.State() == gobreaker.StateOpen {
+		s.log.Debug("Calling overlay fallback method")
+		return fallbackFn()
+	}
+
+	s.log.Debug("Calling overlay base method")
+	result, err := s.circuitBreaker.Execute(func() (interface{}, error) {
+		// TODO(saml) only increment on network specific errors?
+		return baseFn()
+	})
+
+	//nolint:forcetypeassert
+	return result.(T), err
+}
+
+//
+// PolicyLoader interface
+//
+
+func (s *Store) GetPolicySet(ctx context.Context, id namer.ModuleID) (*runtimev1.RunnablePolicySet, error) {
+	// Both `SourceStore` (via `compile.Manager`) and `BinaryStore` implement GetPolicySet
+	return withCircuitBreaker(
+		s,
+		func() (*runtimev1.RunnablePolicySet, error) { return s.basePolicyLoader.GetPolicySet(ctx, id) },
+		func() (*runtimev1.RunnablePolicySet, error) { return s.fallbackPolicyLoader.GetPolicySet(ctx, id) },
+	)
+}
+
+//
+// Store interface methods
+//
+
+func (s *Store) Driver() string {
+	return DriverName
+}
+
+func (s *Store) ListPolicyIDs(ctx context.Context, includeDisabled bool) ([]string, error) {
+	return withCircuitBreaker(
+		s,
+		func() ([]string, error) { return s.baseStore.ListPolicyIDs(ctx, includeDisabled) },
+		func() ([]string, error) { return s.fallbackStore.ListPolicyIDs(ctx, includeDisabled) },
+	)
+}
+
+func (s *Store) ListSchemaIDs(ctx context.Context) ([]string, error) {
+	return withCircuitBreaker(
+		s,
+		func() ([]string, error) { return s.baseStore.ListSchemaIDs(ctx) },
+		func() ([]string, error) { return s.fallbackStore.ListSchemaIDs(ctx) },
+	)
+}
+
+func (s *Store) LoadSchema(ctx context.Context, url string) (io.ReadCloser, error) {
+	return withCircuitBreaker(
+		s,
+		func() (io.ReadCloser, error) { return s.baseStore.LoadSchema(ctx, url) },
+		func() (io.ReadCloser, error) { return s.fallbackStore.LoadSchema(ctx, url) },
+	)
+}
+
+func (s *Store) Reload(ctx context.Context) error {
```

Comment:
Shouldn't we just try to reload both stores in parallel? If any of them is not a `Reloadable` then it would just be a noop. I don't think we should use the circuit breaker here because that would result in only one store being reloaded.

I'd use a [`conc.Pool`](https://pkg.go.dev/github.com/sourcegraph/conc@v0.3.0/pool#Pool) with   `WithContext().WithErrors().WithFirstError()`.